Clear Headers After File Upload to Prevent Request Issues

darwishdev opened the pull request:
This PR fixes a bug where headers set during file uploads (such as `cache-control`, `content-type`, and `x-upsert`) persist after the request, causing errors in subsequent API calls. Specifically, after a successful upload with options like `upsert: true`, the headers remain unchanged, leading to issues when using the storage client for other operations (e.g., listing files).  

#### **Reproduction Steps:**  
1. Upload a file using the `UploadOrUpdateFile` function with options like `{ Upsert: true }`:  
   ```go
   	response, err := s.StorageClient.UploadFile('images', 'image.webp', reader,  storage_go.FileOptions{
		                            ContentType: &req.FileType,
		                            Upsert:      &isUpsert,
	                            })
   ```
2. Try listing files immediately after the upload:  
   ```go
  	response, err := s.StorageClient.ListFiles('images', "", storage_go.FileSearchOptions{
		                            Limit:  int(req.Limit),
		                            Offset: int(req.Offest),
	                            })
   ```
3. The following error occurs:  
   ```
   "body must be object"
   ```
4. If we reset the `storage_client`, the storage client starts working again, indicating that headers were not properly cleared after the upload request.  

#### **Fix:**  
- Ensured all custom headers are cleared after the request completes using `defer`, preventing them from affecting future requests.  

#### **Issue Resolved:**  
- Fixes the `"body must be object"` error when listing files after an upload with options.

darwishdev committed:
clear client header on  storage upload to prevent error when passing options

darwishdev committed:
make sure the delete headers are defered

darwishdev committed:
fix urls

darwishdev committed:
fix content type header

darwishdev committed:
use supabase community urls back to create a pull request

aboozaid commented:
This is a critical issue and this PR needs to be resolved ASAP

aboozaid commented:
As a workaround for now you can use reflect
```
clientValue := reflect.ValueOf(s.storageClient).Elem()
clientTransportField := clientValue.FieldByName("clientTransport")
headerField := clientTransportField.FieldByName("header")
headerPtr := (*http.Header)(unsafe.Pointer(headerField.UnsafeAddr()))
headerPtr.Del("cache-control")
headerPtr.Del("x-upsert")
headerPtr.Set("content-type", "application/json")
```

muratmirgun commented:
Hello I check this issue asap!

Thanks for contributions 🎊

darwishdev commented:
@aboozaid you can use my fork 
[here](https://github.com/darwishdev/supaapi-go)

on this repo i fixed this issue and added the ability to handle different environments for the supabase client including auth and storage clients

aboozaid commented:
@darwishdev I checked your fork but it still has another issue if you have two upload requests at the same time you will get "fatal error: concurrent map writes" error because of both requests try to access the same header at the same time.

Please use sync.Mutex in order to fix that. I return back to my workaround until you solve it :)

```
clientValue := reflect.ValueOf(s.storageClient).Elem()
clientTransportField := clientValue.FieldByName("clientTransport")
headerField := clientTransportField.FieldByName("header")
headerPtr := (*http.Header)(unsafe.Pointer(headerField.UnsafeAddr()))
mu.Lock()
headerPtr.Del("cache-control")
headerPtr.Del("x-upsert")
headerPtr.Set("content-type", "application/json")
mu.Unlock()
```

aboozaid commented:
It seems like the above solution still does not prevent the issue and it has to be solved from the library itself because I use fileOptions
